util/stop-handler/src/lib.rs

```diff
@@ -0,0 +1,64 @@
+use crossbeam_channel::Sender;
+use futures::sync::oneshot;
+use log::error;
+use parking_lot::Mutex;
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+#[derive(Debug)]
+pub enum SignalSender {
+    Future(oneshot::Sender<()>),
+    Crossbeam(Sender<()>),
+}
+
+impl SignalSender {
+    pub fn send(self) {
+        match self {
+            SignalSender::Crossbeam(tx) => {
+                if let Err(e) = tx.send(()) {
+                    error!("handler signal send error {:?}", e);
+                };
+            }
+            SignalSender::Future(tx) => {
+                if let Err(e) = tx.send(()) {
+                    error!("handler signal send error {:?}", e);
+                };
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Handler<T> {
+    signal: SignalSender,
+    thread: JoinHandle<T>,
+}
+
+#[derive(Clone, Debug)]
+pub struct StopHandler<T> {
+    inner: Option<Arc<Mutex<Option<Handler<T>>>>>,
```

Comment:
Why need two `Option`?

Reply:
Usually we call `try_send(&mut self)` in `Drop::drop(&mut self)`,
but we need take ownership of inner structs, `Option` is for that.

the outer `Option`  take ownership for `Arc::try_unwrap`
```
 pub fn try_unwrap(this: Self) -> Result<T, Self>
```

the inner `Option `  take ownership for `oneshot::Sender` and `JoinHandle`

Reply:
Looks weird, better add this as a comment.

Reply:
I think use `Option` is a common technique to take ownership from a `&mut self`